python/dlisio/load.py

```diff
@@ -88,88 +88,209 @@ def load(path, error_handler = None):
-        # files at the current position, nor is it able to precisly report the
-        # underlying tell. Therefore, dlisio has to manually search for the
-        # VRL to determine the right offset in which to open the new filehandle
-        # at.
-        #
-        # Logical files are partitioned by core.findoffsets and it's required
-        # [1] that new logical files always start on a new Visible Record.
-        # Hence, dlisio takes the (approximate) tell at the end of each Logical
-        # File and searches for the VRL to get the exact tell.
-        #
-        # [1] rp66v1, 2.3.6 Record Structure Requirements:
-        #     > ... Visible Records cannot intersect more than one Logical File.
-        while True:
-            if tapemarks: offset -= tifsize
-            stream.seek(offset)
-            if tapemarks: stream = core.open_tif(stream)
-            stream = core.open_rp66(stream)
-
-            explicits, implicits, broken = core.findoffsets(stream, error_handler)
-            hint = rewind(stream.ptell, tapemarks)
-
-            recs  = core.extract(stream, explicits, error_handler)
-            sets  = core.parse_objects(recs, error_handler)
-            pool  = core.pool(sets)
-            fdata = core.findfdata(stream, implicits, error_handler)
-
-            lf = logicalfile(stream, pool, fdata, sul, error_handler)
-            lfs.append(lf)
-
-            if len(broken):
-                # do not attempt to recover or read more logical files
-                # if the error happened in findoffsets
-                # return all logical files we were able to process until now
-                break
-
-            stream = open(path)
-
-            try:
-                stream.seek(hint)
-                core.findvrl(stream, error_handler)
-                offset = stream.ltell
-            except RuntimeError:
-                if stream.eof():
-                    stream.close()
-                    break
-                raise
-
-        return physicalfile(lfs)
+            if parser.logical_eof():
+                parser.close_stream()
+                continue
+
+            if parser.find_SUL():
+                parser.read_SUL()
+
+            if parser.logical_eof():
+                parser.close_stream()
+                continue
+
+            parser.parse_LF()
     except:
-        stream.close()
-        for f in lfs:
-            f.close()
+        parser.close()
         raise
+
+    return physicalfile(parser.lfs)
+
+
+class LFParser:
+    """ Logical Files Parser
+
+    Contains all the internal information required to correctly parse logical
+    files.
+    """
+    def __init__(self, path, is_tif, error_handler):
+        self.error_handler = error_handler
+        self.is_tif = is_tif
+        self.path = path
+
+        self.lfs = []
+        self.sul = None
+        self.stream = None
+
+        self.open_next_at = 0
+        self.data_eof = False
```

Comment:
Renamed to data_end, but I am against splitting this, because it means too many different things
- physical file EOF
- file truncated
- file truncated and padded which is discovered by findoffsets
- file not truncated but padded which is discovered by lfp

And all of them mean that we no longer have valid data to process.